File: go/gin/internal/modules/client/dto/create_client_dto.go
package dto

import (
	"strings"
	"tcc/internal/common/helper"
	"time"
)

type CreateClientDto struct {
	Nome            string `json:"nome" validate:"required,min=3"`
	CPF             string `json:"cpf" validate:"required,cpf"`
	Telefone        string `json:"telefone" validate:"required"`
	DataDeNascimento string `json:"dataDeNascimento" validate:"required,date_format"`
	Email           string `json:"email" validate:"required,email"`
	Senha           string `json:"senha" validate:"required,min=8,number"`
}

func (dto *CreateClientDto) Transform() {
	// Nome
	dto.Nome = strings.TrimSpace(dto.Nome)
	dto.Nome = helper.RemoveAccents(dto.Nome)
	dto.Nome = strings.ToUpper(dto.Nome)

	// CPF
	dto.CPF = helper.CleanDigits(dto.CPF)

	// Telefone
	dto.Telefone = helper.CleanDigits(dto.Telefone)

	// DataDeNascimento
	t, err := time.Parse("2001-11-09", dto.DataDeNascimento)
	if err == nil {
		dto.DataDeNascimento = t.Format("2001-11-09")
	}
}

File: rust/axum/src/modules/client/dto/create_client_dto.rs
use chrono::NaiveDate;
use serde::{Deserialize, Serialize};
use validator::Validate;
use crate::common::helper::{clean_digits::clean_digits, remove_accents::remove_accents};
use crate::common::validator::validate_cpf;

#[derive(Debug, Deserialize, Serialize, Validate)]
pub struct CreateClientDto {
    #[validate(length(min = 3, message="O campo nome deve ter no mínimo 3 caracteres."))]
    pub nome: String,
    #[validate(custom(function = "validate_cpf"))]
    pub cpf: String,
    #[validate(length(min = 1,message = "O campo telefone é obrigatório."))]
    pub telefone: String,
    #[serde(rename = "dataDeNascimento", with = "naive_date_format")]
    pub data_de_nascimento: NaiveDate,
    #[validate(email(message = "O campo email deve ser um email."))]
    pub email: String,
    #[validate(length(min = 8, message = "O campo senha deve ter no mínimo 8 caracteres."))]
    pub senha: String,
}

impl CreateClientDto {
    pub fn transform(&mut self) {
        // Nome
        self.nome = remove_accents(&self.nome.trim().to_uppercase());

        // CPF
        self.cpf = clean_digits(&self.cpf);

        // Telefone
        self.telefone = clean_digits(&self.telefone);
    }
}


mod naive_date_format {
    use chrono::NaiveDate;
    use serde::{self, Deserialize, Deserializer, Serializer};

    const FORMAT: &str = "%Y-%m-%d";

    pub fn serialize<S>(date: &NaiveDate, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let s = date.format(FORMAT).to_string();
        serializer.serialize_str(&s)
    }

    pub fn deserialize<'de, D>(deserializer: D) -> Result<NaiveDate, D::Error>
    where
        D: Deserializer<'de>,
    {
        let s = String::deserialize(deserializer)?;
        NaiveDate::parse_from_str(&s, FORMAT).map_err(|_| {
            serde::de::Error::custom("O campo dataDeNascimento deve ser uma data válida.")
        })
    }
}